File: java/src/main/java/org/libpetri/core/Out.java
package org.libpetri.core;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Output specification with explicit split semantics.
 * Supports composite structures (XOR of ANDs, AND of XORs, etc.)
 *
 * <p>This is a sealed hierarchy enabling pattern matching and formal analysis.
 * The structure encodes the contract that the runtime enforces:
 * <ul>
 *   <li>{@link And} - ALL children must receive tokens</li>
 *   <li>{@link Xor} - EXACTLY ONE child receives token</li>
 *   <li>{@link Place} - Leaf node representing a single output place</li>
 *   <li>{@link Timeout} - Timeout branch that activates if action exceeds duration</li>
 * </ul>
 *
 * <h3>Usage Examples</h3>
 * <pre>{@code
 * // Simple XOR (flat)
 * Out.xor(successPlace, errorPlace)
 *
 * // Simple AND (flat)
 * Out.and(p1, p2, p3)
 *
 * // XOR of ANDs (choose one branch, produce to all in branch)
 * Out.xor(
 *     Out.and(headerPlace, bodyPlace),     // Branch A
 *     Out.and(errorPlace, logPlace)        // Branch B
 * )
 *
 * // AND of XORs (multiple independent choices)
 * Out.and(
 *     Out.xor(formatA, formatB),           // Choice 1
 *     Out.xor(destA, destB)                // Choice 2
 * )
 * }</pre>
 *
 * @see In for input specifications
 */
public sealed interface Out permits Out.And, Out.Xor, Out.Place, Out.Timeout, Out.ForwardInput {

    /**
     * AND-split: ALL children must receive tokens.
     * Validation fails if any child is not satisfied.
     */
    record And(List<Out> children) implements Out {
        public And {
            if (children == null) {
                throw new IllegalArgumentException("AND children cannot be null");
            }
            if (children.isEmpty()) {
                throw new IllegalArgumentException("AND requires at least 1 child");
            }
            if (children.stream().anyMatch(c -> c == null)) {
                throw new IllegalArgumentException("AND children cannot contain null elements");
            }
            children = List.copyOf(children);
        }
    }

    /**
     * XOR-split: EXACTLY ONE child receives token.
     * Validation fails if zero or more than one child is satisfied.
     */
    record Xor(List<Out> children) implements Out {
        public Xor {
            if (children == null) {
                throw new IllegalArgumentException("XOR children cannot be null");
            }
            if (children.size() < 2) {
                throw new IllegalArgumentException("XOR requires at least 2 children");
            }
            if (children.stream().anyMatch(c -> c == null)) {
                throw new IllegalArgumentException("XOR children cannot contain null elements");
            }
            children = List.copyOf(children);
        }
    }

    /**
     * Leaf node: a single output place.
     */
    record Place(org.libpetri.core.Place<?> place) implements Out {}

    /**
     * Timeout branch - activates if action exceeds duration.
     *
     * <p>When an action doesn't complete within the specified duration,
     * the action is cancelled and tokens are produced to the child output(s).
     *
     * <p>This is part of the output structure, not a separate concern.
     * The executor interprets Timeout nodes and enforces them.
     *
     * <h3>Usage Example</h3>
     * <pre>{@code
     * .outputs(Out.xor(
     *     successPlace,
     *     Out.timeout(Duration.ofSeconds(5), timeoutPlace),  // Timeout branch
     *     errorPlace
     * ))
     * }</pre>
     */
    record Timeout(Duration after, Out child) implements Out {
        public Timeout {
            if (after == null || after.isNegative() || after.isZero()) {
                throw new IllegalArgumentException("Timeout must be positive: " + after);
            }
            if (child == null) {
                throw new IllegalArgumentException("Timeout child cannot be null");
            }
        }
    }

    /**
     * Forward a consumed input token to an output place.
     * Used in timeout branches to retry with original input data.
     *
     * <p>The 'from' place must be an input place of the transition.
     * The value consumed from 'from' at transition start is produced to 'to' on timeout.
     *
     * <h3>Usage Example</h3>
     * <pre>{@code
     * .inputs(In.one(queryPlace))
     * .outputs(Out.xor(
     *     resultPlace,
     *     Out.timeout(Duration.ofSeconds(10),
     *         Out.forwardInput(queryPlace, retryPlace))
     * ))
     * }</pre>
     */
    record ForwardInput(org.libpetri.core.Place<?> from, org.libpetri.core.Place<?> to) implements Out {
        public ForwardInput {
            if (from == null || to == null) {
                throw new IllegalArgumentException("ForwardInput places cannot be null");
            }
        }
    }

    // ==================== Factory Methods ====================

    /**
     * Creates an AND-split with the given children.
     * All children must be satisfied for validation to pass.
     *
     * @param children the child output specs
     * @return AND output spec
     */
    static And and(Out... children) {
        return new And(List.of(children));
    }

    /**
     * Creates an AND-split with the given places.
     * All places must receive tokens for validation to pass.
     *
     * @param places the output places
     * @return AND output spec
     */
    static And and(org.libpetri.core.Place<?>... places) {
        return new And(Arrays.stream(places).<Out>map(Out.Place::new).toList());
    }

    /**
     * Creates a XOR-split with the given children.
     * Exactly one child must be satisfied for validation to pass.
     *
     * @param children the child output specs
     * @return XOR output spec
     */
    static Xor xor(Out... children) {
        return new Xor(List.of(children));
    }

    /**
     * Creates a XOR-split with the given places.
     * Exactly one place must receive a token for validation to pass.
     *
     * @param places the output places
     * @return XOR output spec
     */
    static Xor xor(org.libpetri.core.Place<?>... places) {
        return new Xor(Arrays.stream(places).<Out>map(Out.Place::new).toList());
    }

    /**
     * Creates a leaf output spec for a single place.
     *
     * @param p the output place
     * @return Place output spec
     */
    static Place place(org.libpetri.core.Place<?> p) {
        return new Place(p);
    }

    /**
     * Creates a timeout output spec with the given duration and child output.
     *
     * <p>If the action doesn't complete within the specified duration,
     * the action is cancelled and tokens are produced to the child output(s).
     *
     * @param after the timeout duration
     * @param child the output spec to use on timeout
     * @return Timeout output spec
     */
    static Timeout timeout(Duration after, Out child) {
        return new Timeout(after, child);
    }

    /**
     * Creates a timeout output spec with the given duration and place.
     *
     * <p>If the action doesn't complete within the specified duration,
     * the action is cancelled and a token is produced to the specified place.
     *
     * @param after the timeout duration
     * @param p the place to produce to on timeout
     * @return Timeout output spec
     */
    static Timeout timeout(Duration after, org.libpetri.core.Place<?> p) {
        return new Timeout(after, place(p));
    }

    /**
     * Creates a forward-input output spec for timeout branches.
     *
     * <p>When the transition times out, the value consumed from the 'from' input place
     * is produced to the 'to' output place. This enables retry patterns where the
     * original input data needs to be forwarded to a retry handler.
     *
     * @param from the input place whose consumed value will be forwarded
     * @param to the output place to produce the forwarded value to
     * @return ForwardInput output spec
     */
    static ForwardInput forwardInput(org.libpetri.core.Place<?> from, org.libpetri.core.Place<?> to) {
        return new ForwardInput(from, to);
    }

    // ==================== Instance Methods ====================

    /**
     * Collects all leaf places from this output spec (flattened).
     * Useful for TransitionContext validation of allowed outputs.
     *
     * @return unmodifiable set of all places in this spec
     */
    default Set<org.libpetri.core.Place<?>> allPlaces() {
        return switch (this) {
            case Place p -> Set.of(p.place());
            case ForwardInput f -> Set.of(f.to());  // Output place only
            case And a -> a.children().stream()
                .flatMap(c -> c.allPlaces().stream())
                .collect(Collectors.toUnmodifiableSet());
            case Xor x -> x.children().stream()
                .flatMap(c -> c.allPlaces().stream())
                .collect(Collectors.toUnmodifiableSet());
            case Timeout t -> t.child().allPlaces();
        };
    }

    /**
     * Enumerates all possible output branches for structural analysis.
     *
     * <p>This method is used by the {@code StateClassGraph} analyzer to expand
     * XOR outputs into virtual transitions (one per branch). Each branch represents
     * a distinct possible outcome of firing the transition.
     *
     * <ul>
     *   <li>AND = single branch containing all child places (Cartesian product)</li>
     *   <li>XOR = one branch per alternative child</li>
     *   <li>Nested = Cartesian product for AND, union for XOR</li>
     * </ul>
     *
     * <h3>Examples</h3>
     * <pre>{@code
     * Out.and(a, b).enumerateBranches()           // [{a, b}]
     * Out.xor(a, b).enumerateBranches()           // [{a}, {b}]
     * Out.xor(Out.and(a,b), Out.and(c,d))         // [{a,b}, {c,d}]
     * Out.and(Out.xor(a,b), Out.xor(c,d))         // [{a,c}, {a,d}, {b,c}, {b,d}]
     * }</pre>
     *
     * @return list of branches, where each branch is a set of places
     */
    default List<Set<org.libpetri.core.Place<?>>> enumerateBranches() {
        return switch (this) {
            case Place p -> List.of(Set.of(p.place()));
            case ForwardInput f -> List.of(Set.<org.libpetri.core.Place<?>>of(f.to()));

            case And and -> {
                // Cartesian product: each combination of children
                List<Set<org.libpetri.core.Place<?>>> result = List.of(Set.of());
                for (Out child : and.children()) {
                    result = crossProduct(result, child.enumerateBranches());
                }
                yield result;
            }

            case Xor xor -> {
                // Union: each child is a separate branch
                var result = new ArrayList<Set<org.libpetri.core.Place<?>>>();
                for (Out child : xor.children()) {
                    result.addAll(child.enumerateBranches());
                }
                yield result;
            }

            case Timeout t -> t.child().enumerateBranches();
        };
    }

    /**
     * Computes the Cartesian product of two branch lists.
     * Each resulting branch is the union of one branch from each input list.
     */
    private static List<Set<org.libpetri.core.Place<?>>> crossProduct(
            List<Set<org.libpetri.core.Place<?>>> a,
            List<Set<org.libpetri.core.Place<?>>> b) {
        var result = new ArrayList<Set<org.libpetri.core.Place<?>>>();
        for (var setA : a) {
            for (var setB : b) {
                var merged = new HashSet<org.libpetri.core.Place<?>>(setA);
                merged.addAll(setB);
                result.add(Set.copyOf(merged));
            }
        }
        return result;
    }
}

File: typescript/src/core/out.ts
import type { Place } from './place.js';

/**
 * Output specification with explicit split semantics.
 * Supports composite structures (XOR of ANDs, AND of XORs, etc.)
 *
 * - And: ALL children must receive tokens
 * - Xor: EXACTLY ONE child receives token
 * - Place: Leaf node representing a single output place
 * - Timeout: Timeout branch that activates if action exceeds duration
 * - ForwardInput: Forward consumed input to output on timeout
 *
 * A spec names **places, not counts**. Validation ([IO-015]) compares the SET of
 * places an action wrote against the branches' claims, so an action that deposits
 * several tokens into one named place is accepted — and every analysis that
 * enumerates branches ({@link enumerateBranches}: the state-class graph, the SMT
 * encoding, the ν fragment check) models exactly one token per named place. Such a
 * firing therefore does more than the analyses explore, in the direction that can
 * make a `proven` false. The executors report it once per transition as a `WARN`
 * log-message ([IO-016]); a net meant to be verified should produce one token per
 * named place and express multiplicity in its topology.
 */
export type Out = OutAnd | OutXor | OutPlace | OutTimeout | OutForwardInput;

export interface OutAnd {
  readonly type: 'and';
  readonly children: readonly Out[];
}

export interface OutXor {
  readonly type: 'xor';
  readonly children: readonly Out[];
}

export interface OutPlace {
  readonly type: 'place';
  readonly place: Place<any>;
}

export interface OutTimeout {
  readonly type: 'timeout';
  /** Timeout duration in milliseconds. */
  readonly afterMs: number;
  readonly child: Out;
}

export interface OutForwardInput {
  readonly type: 'forward-input';
  readonly from: Place<any>;
  readonly to: Place<any>;
}

// ==================== Factory Functions ====================

/**
 * AND-split: all children must receive tokens.
 *
 * @example
 * ```ts
 * // AND of XOR branches: one of (A,B) AND one of (C,D)
 * and(xorPlaces(placeA, placeB), xorPlaces(placeC, placeD))
 *
 * // AND with a fixed place + XOR branch
 * and(outPlace(always), xorPlaces(left, right))
 * ```
 */
export function and(...children: Out[]): OutAnd {
  if (children.length === 0) {
    throw new Error('AND requires at least 1 child');
  }
  return { type: 'and', children };
}

/** AND-split from places: all places must receive tokens. */
export function andPlaces(...places: Place<any>[]): OutAnd {
  return and(...places.map(outPlace));
}

/** XOR-split: exactly one child receives token. */
export function xor(...children: Out[]): OutXor {
  if (children.length < 2) {
    throw new Error('XOR requires at least 2 children');
  }
  return { type: 'xor', children };
}

/** XOR-split from places: exactly one place receives token. */
export function xorPlaces(...places: Place<any>[]): OutXor {
  return xor(...places.map(outPlace));
}

/** Leaf output spec for a single place. */
export function outPlace(p: Place<any>): OutPlace {
  return { type: 'place', place: p };
}

/** Timeout output: activates if action exceeds duration. */
export function timeout(afterMs: number, child: Out): OutTimeout {
  if (afterMs <= 0) {
    throw new Error(`Timeout must be positive: ${afterMs}`);
  }
  return { type: 'timeout', afterMs, child };
}

/** Timeout output pointing to a single place. */
export function timeoutPlace(afterMs: number, p: Place<any>): OutTimeout {
  return timeout(afterMs, outPlace(p));
}

/** Forward consumed input value to output place on timeout. */
export function forwardInput(from: Place<any>, to: Place<any>): OutForwardInput {
  return { type: 'forward-input', from, to };
}

// ==================== Helper Functions ====================

/** Collects all leaf places from this output spec (flattened). */
export function allPlaces(out: Out): Set<Place<any>> {
  const result = new Set<Place<any>>();
  collectPlaces(out, result);
  return result;
}

function collectPlaces(out: Out, result: Set<Place<any>>): void {
  switch (out.type) {
    case 'place':
      result.add(out.place);
      break;
    case 'forward-input':
      result.add(out.to);
      break;
    case 'and':
    case 'xor':
      for (const child of out.children) {
        collectPlaces(child, result);
      }
      break;
    case 'timeout':
      collectPlaces(out.child, result);
      break;
  }
}

/**
 * Enumerates all possible output branches for structural analysis ([IO-016]).
 *
 * - AND = single branch containing all child places (Cartesian product)
 * - XOR = one branch per alternative child
 * - Nested = Cartesian product for AND, union for XOR
 *
 * A branch is a **set** of places: it says which places receive a token, not how
 * many tokens each receives. Analyses built on it (the state-class graph's virtual
 * transitions, the flattener's post vectors, the ν fragment check) deposit one token
 * per place of the chosen branch. An action that writes `n > 1` tokens to a place
 * its branch names once is accepted by [IO-015] but is outside what those analyses
 * explore — a sound under-approximation for safety only when it never happens, which
 * is why the executors warn about it ([IO-016] AC4).
 */
export function enumerateBranches(out: Out): ReadonlyArray<ReadonlySet<Place<any>>> {
  switch (out.type) {
    case 'place':
      return [new Set([out.place])];

    case 'forward-input':
      return [new Set<Place<any>>([out.to])];

    case 'and': {
      let result: Set<Place<any>>[] = [new Set()];
      for (const child of out.children) {
        result = crossProduct(result, enumerateBranches(child) as Set<Place<any>>[]);
      }
      return result;
    }

    case 'xor': {
      const result: Set<Place<any>>[] = [];
      for (const child of out.children) {
        result.push(...(enumerateBranches(child) as Set<Place<any>>[]));
      }
      return result;
    }

    case 'timeout':
      return enumerateBranches(out.child);
  }
}

function crossProduct(
  a: Set<Place<any>>[],
  b: ReadonlyArray<ReadonlySet<Place<any>>>,
): Set<Place<any>>[] {
  const result: Set<Place<any>>[] = [];
  for (const setA of a) {
    for (const setB of b) {
      const merged = new Set<Place<any>>(setA);
      for (const p of setB) merged.add(p);
      result.push(merged);
    }
  }
  return result;
}
